apollo-router-core/src/cache.rs

```diff
@@ -0,0 +1,176 @@
+        // the delegate get() calls into v8 for processing of the plan.
+        // This would block all other get() requests for a potentially
+        // long time.
+        // Alternatively, if we don't hold the lock, there is a risk
+        // that we will do the work multiple times. This is also
+        // sub-optimal.
+
+        // To work around this, we keep a list of keys we are currently
+        // processing in the delegate. If we try to get a key on this
+        // list, we block and wait for it to complete and then retry.
+        //
+        // This is more complex than either of the two simple
+        // alternatives but succeeds in providing a mechanism where each
+        // client only waits for uncached QueryPlans they are going to
+        // use AND avoids generating the plan multiple times.
+
+        let mut locked_wait_map = self.wait_map.lock().await;
+
+        // We must only drop the locked cache after we have locked the
+        // wait map. Otherwise,we might get a race that causes us to
+        // miss a broadcast.
+        drop(locked_cache);
+
+        match locked_wait_map.get_mut(&key) {
+            Some(waiter) => {
+                // Register interest in key
+                let mut receiver = waiter.subscribe();
+                drop(locked_wait_map);
+                // Our use case is very specific, so we are sure
+                // that we won't get any errors here.
+                let (recv_key, recv_plan) = receiver.recv().await.expect(
+                    "the sender won't ever be dropped before all the receivers finish; qed",
+                );
+                debug_assert_eq!(recv_key, key);
+                recv_plan
+            }
+            None => {
+                let (tx, _rx) = broadcast::channel(1);
+                locked_wait_map.insert(key.clone(), tx.clone());
+                drop(locked_wait_map);
+                // This is the potentially high duration operation where we ask our resolver to
+                // resolve the key (retrieve a value) for us
+                // No cache locks are held here
+                let value = self.resolver.retrieve(key.clone()).await;
+                // Update our cache
+                let mut locked_cache = self.cached.lock().await;
+                locked_cache.put(key.clone(), value.clone());
+                // Update our wait list
+                let mut locked_wait_map = self.wait_map.lock().await;
+                locked_wait_map.remove(&key);
+                // Let our waiters know
+                let broadcast_value = value.clone();
+                // Our use case is very specific, so we are sure that
+                // we won't get any errors here.
+                tokio::task::spawn_blocking(move || {
+                    tx.send((key, broadcast_value))
+                        .expect("there is always at least one receiver alive, the _rx guard; qed")
+                })
+                .await?;
```

Comment:
This could fail only if the task is aborted (not possible with spawn blocking) of if the code insides panic. It doesn't seem like it could panic so you can put the same expect here and drop one of your error variant.

Reply:
Or:

You can ignore the result entirely because it doesn't impact the working of this function and this request.

(and just log it with log::debug)

Reply:
aside:
If the send fails, then a number of tasks may be waiting for the broadcast to arrive. Doesn't that mean that our expect at line 92 will trigger?

Reply:
send() can only fail if there are no receivers expecting it: https://docs.rs/tokio/1.14.0/tokio/sync/broadcast/error/struct.SendError.html

The only case where the receiver might wait indefinitely is if we do not send and just return, because the wait map contains another clone of the sender, so the receiver won't get a notification that the broadcast channel is closed

Reply:
Resolved in: https://github.com/apollographql/router/pull/235/commits/84a043d61076bfa66746868501519addb17f2193